metals/src/main/scala/scala/meta/internal/metals/ReferenceProvider.scala

```diff
@@ -334,57 +373,55 @@ final class ReferenceProvider(
     }
   }
 
-  private def pcReferences(
-      path: AbsolutePath,
-      params: ReferenceParams,
-      adjustLocation: AdjustRange,
-  ): Future[List[ReferencesResult]] = {
-    compilers.references(params, EmptyCancelToken, adjustLocation).flatMap {
-      foundRefs =>
-        pcReferences(
-          path,
-          foundRefs.map(_.symbol),
-          includeDeclaration = params.getContext().isIncludeDeclaration(),
-          filterTargetFiles = _ != path,
-          adjustLocation,
-        ).map(_ ++ foundRefs)
-    }
-  }
-
   private def pcReferences(
       path: AbsolutePath,
       symbols: List[String],
       includeDeclaration: Boolean,
-      filterTargetFiles: AbsolutePath => Boolean,
       adjustLocation: AdjustRange,
   ): Future[List[ReferencesResult]] = {
     val visited = mutable.Set[AbsolutePath]()
-    val results = for {
-      buildTarget <- buildTargets.inverseSources(path).toList
-      _ = visited.clear()
-      symbol <- symbols
-      name = nameFromSymbol(symbol)
-      pathsMap = pathsForName(buildTarget, name)
-      id <- pathsMap.keySet
-      searchFiles = pathsMap(id)
-        .filter(searchFile =>
-          filterTargetFiles(searchFile) && !visited(searchFile)
-        )
-        .distinct
-      if searchFiles.nonEmpty
-    } yield {
-      visited ++= searchFiles
-      () =>
-        compilers.references(
-          id,
-          searchFiles,
-          includeDeclaration,
-          symbol,
-          adjustLocation,
-        )
-    }
-    val maxPcsNumber = Runtime.getRuntime().availableProcessors() / 2
-    executeBatched(results, maxPcsNumber).map(_.flatten)
+    val names = symbols.map(nameFromSymbol(_)).toSet
+    val pathsWithId =
+      for {
+        buildTarget <- buildTargets.inverseSources(path).toList
+        name <- names
+        pathsMap = pathsForName(buildTarget, name)
+        id <- pathsMap.keySet
+      } yield {
+        val searchFiles = pathsMap(id).filter { searchFile =>
+          val indexedFile = index.get(searchFile.toNIO)
+          (indexedFile.isEmpty || indexedFile.get.isStale) && !visited(
+            searchFile
+          )
+        }.distinct
+        visited ++= searchFiles
+        (id -> searchFiles)
+      }
+
+    val results = pathsWithId
```

Comment:
```suggestion
    val lazyResults = pathsWithId
```
minor, so that we actually see that this is functions being retrned